core/src/main/java/org/elasticsearch/index/translog/Translog.java

```diff
@@ -409,20 +411,40 @@ TranslogWriter createWriter(long fileGeneration) throws IOException {
     public Location add(final Operation operation) throws IOException {
         final ReleasableBytesStreamOutput out = new ReleasableBytesStreamOutput(bigArrays);
         try {
-            final BufferedChecksumStreamOutput checksumStreamOutput = new BufferedChecksumStreamOutput(out);
             final long start = out.position();
             out.skip(Integer.BYTES);
-            writeOperationNoSize(checksumStreamOutput, operation);
+            writeOperationNoSize(new BufferedChecksumStreamOutput(out), operation);
             final long end = out.position();
             final int operationSize = (int) (end - Integer.BYTES - start);
             out.seek(start);
             out.writeInt(operationSize);
             out.seek(end);
             final ReleasablePagedBytesReference bytes = out.bytes();
+            final Location location;
+            final boolean shouldRollGeneration;
             try (ReleasableLock ignored = readLock.acquire()) {
                 ensureOpen();
-                return current.add(bytes, operation.seqNo());
+                location = current.add(bytes, operation.seqNo());
+                // check if we should roll under the read lock
+                shouldRollGeneration =
+                        shouldRollGeneration() && rollingGeneration.compareAndSet(false, true);
             }
+            if (shouldRollGeneration) {
+                try (ReleasableLock ignored = writeLock.acquire()) {
+                    /*
+                     * We have to check the condition again lest we could roll twice if another
+                     * thread committed the translog (which rolls the generation) between us
+                     * releasing the read lock and acquiring the write lock.
+                     */
+                    if (shouldRollGeneration()) {
+                        this.rollGeneration();
+                    }
+                } finally {
+                    final boolean wasRolling = rollingGeneration.getAndSet(false);
```

Comment:
we have a similar logic in `IndexShard#maybeFlush()` which I think we should extract into a utility class and we can test separately and reuse in both places. The logic looks pretty much the same and given how error prone that is we should really make sure we isolate and reuse.